### PR TITLE
janeStreet: 0.9.0 -> 0.10.0 [WIP]

### PR DESCRIPTION
**Pull request by mt-caret**
###### Motivation for this change

Attempt to update janeStreet packages.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/ocaml-modules/janestreet/default.nix [hunk 1]
```diff
@@ -34,8 +32,7 @@ rec {
     meta.description = "OCaml AST used by Jane Street ppx rewriters";
   } // (if lib.versionAtLeast ocaml.version "4.06"
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/default.nix]: This test should be dropped.

File: pkgs/development/ocaml-modules/janestreet/default.nix [hunk 2]
```diff
@@ -369,95 +371,98 @@ rec {
...
-    hash = "02clpz3xv3i5avzifwalylb9gfxzpgnr8bnlfsjixxfk2m7kvsj2";
+    hash = "1snkr944l3a627k23yh8f0lr900dpg2aym2l59fpp75s29pyk5md";
     propagatedBuildInputs = [ core_extended async ];
     meta = {
       description = "Shell helpers for Async";
     };
   };
 
+  # Broken due to ctypes build failure (upgrading ctypes may possibly work, but
+  # opens a world of pain)
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/default.nix]: See  #35418

File: pkgs/development/ocaml-modules/janestreet/default.nix [hunk 3]
```diff
@@ -505,111 +510,119 @@ rec {
...
   };
 
   posixat = janePackage {
     name = "posixat";
-    hash = "0ak93dyzi6sc6gb0j07fj85b24d8bv6g2hm7jj5xwb39kjwh51jl";
+    hash = "0dmjzbpbmzl94h4c1gk6k75wglnvk1kqcm4zs4nb9hy2ja8ldl9x";
     propagatedBuildInputs = [ ppx_sexp_conv ];
     meta.description = "Binding to the posix *at functions";
-    meta.broken = lib.versionAtLeast ocaml.version "4.05";
+    meta.broken = !lib.versionAtLeast ocaml.version "4.05";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/default.nix]: This line should be removed.

File: pkgs/development/ocaml-modules/janestreet/default.nix [hunk 4]
```diff
@@ -505,111 +510,119 @@ rec {
...
     name = "topological_sort";
-    hash = "1d64fyq0clsgham9p1f5rk01z8pxalglp92xmqw2iznyw0vxhvsy";
+    hash = "08w1dx30frj2bxxk8djl23cd43sassjkrmissyhagn9fmq2l904m";
     propagatedBuildInputs = [ core_kernel ];
     meta.description = "Topological sort algorithm";
   };
 
   typerep_extended = janePackage {
     name = "typerep_extended";
+    version = "0.9.0";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/default.nix]: Does it make sense to keep this package?

**Review by vbgl — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/janestreet/default.nix (hunk 2 above)
Comment (mt-caret) [reply pkgs/development/ocaml-modules/janestreet/default.nix]: I'll fix this when the PR gets merged.

**Comment by vbgl** [comment]
`ctypes` has been updated. Can you please rebase?

**Comment by mt-caret** [comment]
Hmm, did I mess up rebase? :cry:

**Comment by Ericson2314** [comment]
hehe yes. Use `git rebase -i` and try again.

**Comment by mt-caret** [comment]
I think I got it right this time.
@vbgl Sorry for the delay.

File: pkgs/development/ocaml-modules/janestreet/janePackage.nix [hunk 5]
```diff
@@ -1,6 +1,6 @@
 { stdenv, fetchFromGitHub, ocaml, jbuilder, findlib }:
 
-{ name, version ? "0.9.0", buildInputs ? [], hash, meta, ...}@args:
+{ name, version ? "0.10.0", buildInputs ? [], hash, meta, ...}@args:
 
 if !stdenv.lib.versionAtLeast ocaml.version "4.03"
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/janePackage.nix]: Did you test with 4.03 ? I think 4.04 or above is required.

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocamlPackages.bap

**Comment by vbgl** [comment]
This PR breaks `bad`. That’s an annoying regression.

**Comment by mt-caret** [comment]
I guess the solution is to keep the 0.9.0 versions around (under a different name like janeStreet_0_9_0, perhaps) and have use that version for all of bap's dependencies (and all of their dependencies, and so on). I have no idea how one will go about doing that or if that's even possible.

I've tried to inject the 0.9.0 versions of core_kernel and ppx_jane for bap, but of course that ends up with errors about other dependencies dependent on core_kernel/ppx_jane/etc. 0.10.0.

How should we proceed?

**Comment by vbgl** [comment]
Maybe ask upstream (`bap`) plans about compatibility with `core_kernel-0.10.0`?

**Comment by vbgl** [comment]
Possible solution: keep the old `janestreet/default.nix` file around (under a new name, fixing a few version strings that used to be implicit), import it (as before) under a specific attribute (say `janeStreet_0_9_0`) and use this attribute to feed the `bap` derivation (as in `inherit (janeStreet_0_9_0) core_kernel ppx_jane parsexp`).

This shouldn’t introduce regression and allow to update the whole bunch of janestreet libraries.

**Comment by mt-caret** [comment]
I think just keeping the old janestreet/default.nix works as a temporary workaround while we wait for bap upstream to upgrade to 0.10.0 (https://github.com/BinaryAnalysisPlatform/bap/issues/786). As a bonus, if any other regressions pop up we can also fix them with janestreet/old.nix pretty easily.

**Comment by vbgl** [comment]
@GrahamcOfBorg build patdiff

**Comment by ivg** [comment]
FYI, BAP has a half-a-year release cycle, so we will unlikely upgrade to the core.v0.10 earlier than that. Looping in @maurer.

**Comment by vbgl** [comment]
I guess `patdiff` needs to be updated too.

**Comment by mt-caret** [comment]
@vbgl Done.

**Review by vbgl — APPROVED** [review]

**Comment by mt-caret** [comment]
:tada:

**Pull request merged**